File: app/src/main/java/io/pslab/communication/PacketHandler.java
package io.pslab.communication;

import android.util.Log;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.BufferedReader;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.ByteBuffer;
import java.nio.ByteOrder;
import java.nio.charset.StandardCharsets;
import java.util.Arrays;

import io.pslab.interfaces.HttpCallback;
import io.pslab.others.ScienceLabCommon;

/**
 * Created by viveksb007 on 28/3/17.
 */

public class PacketHandler {

    private static final String TAG = "PacketHandler";
    private final int BUFSIZE = 10000;
    private byte[] buffer = new byte[BUFSIZE];
    private boolean loadBurst, connected;
    private int inputQueueSize = 0, BAUD = 1000000;
    private CommunicationHandler mCommunicationHandler = null;
    public static String version = "";
    private CommandsProto mCommandsProto;
    private int timeout = 500, VERSION_STRING_LENGTH = 15;
    ByteBuffer burstBuffer = ByteBuffer.allocate(2000);
    private HttpAsyncTask httpAsyncTask;

    public PacketHandler(int timeout, CommunicationHandler communicationHandler) {
        this.loadBurst = false;
        this.connected = false;
        this.timeout = timeout;
        this.mCommandsProto = new CommandsProto();
        this.mCommunicationHandler = communicationHandler;
        connected = (mCommunicationHandler.isConnected() || ScienceLabCommon.isWifiConnected());
    }

    public boolean isConnected() {
        connected = (mCommunicationHandler.isConnected() || ScienceLabCommon.isWifiConnected());
        return connected;
    }

    public String getVersion() {
        try {
            sendByte(mCommandsProto.COMMON);
            sendByte(mCommandsProto.GET_VERSION);
            // Read "<PSLAB Version String>\n"
            commonRead(VERSION_STRING_LENGTH + 1);
            // Only use first line, just like in the Python implementation.
            version = new BufferedReader(
                    new InputStreamReader(
                            new ByteArrayInputStream(buffer, 0, VERSION_STRING_LENGTH),
                            StandardCharsets.UTF_8))
                    .readLine();
        } catch (IOException e) {
            Log.e("Error in Communication", e.toString());
        }
        return version;
    }

    public void sendByte(int val) throws IOException {
        if (!connected) {
            throw new IOException("Device not connected");
        }
        if (!loadBurst) {
            try {
                commonWrite(new byte[]{(byte) (val & 0xff)});
            } catch (IOException | NullPointerException e) {
                Log.e("Error in sending byte", e.toString());
                e.printStackTrace();
            }
        } else {
            burstBuffer.put((byte) (val & 0xff));
        }
    }

    public void sendInt(int val) throws IOException {
        if (!connected) {
            throw new IOException("Device not connected");
        }
        if (!loadBurst) {
            try {
                commonWrite(new byte[]{(byte) (val & 0xff), (byte) ((val >> 8) & 0xff)});
            } catch (IOException e) {
                Log.e("Error in sending int", e.toString());
                e.printStackTrace();
            }
        } else {
            burstBuffer.put(new byte[]{(byte) (val & 0xff), (byte) ((val >> 8) & 0xff)});
        }
    }

    public int getAcknowledgement() {
        /*
        fetches the response byte
        1 SUCCESS
        2 ARGUMENT_ERROR
        3 FAILED
        used as a handshake
        */
        if (loadBurst) {
            inputQueueSize++;
            return 1;
        } else {
            try {
                commonRead(1);
                return buffer[0];
            } catch (IOException | NullPointerException e) {
                e.printStackTrace();
                return 3;
            }
        }
    }

    public byte getByte() {
        try {
            int numByteRead = commonRead(1);
            if (numByteRead == 1) {
                return buffer[0];
            } else {
                Log.e(TAG, "Error in reading byte");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return -1;
    }

    int getVoltageSummation() {
        try {
            // Note : bytesToBeRead has to be +1 than the requirement
            int numByteRead = commonRead(3);
            if (numByteRead == 3) {
                return (buffer[0] & 0xff) | ((buffer[1] << 8) & 0xff00);
            } else {
                Log.e(TAG, "Error in reading byte");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return -1;
    }

    public int getInt() {
        try {
            int numByteRead = commonRead(2);
            if (numByteRead == 2) {
                // LSB is read first
                return (buffer[0] & 0xff) | ((buffer[1] << 8) & 0xff00);
            } else {
                Log.e(TAG, "Error in reading byte");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return -1;
    }

    public long getLong() {
        try {
            int numByteRead = commonRead(4);
            if (numByteRead == 4) {
                // C++ has long of 4-bytes but in Java int has 4-bytes
                // refer "https://stackoverflow.com/questions/7619058/convert-a-byte-array-to-integer-in-java-and-vice-versa" for Endian
                return ByteBuffer.wrap(Arrays.copyOfRange(buffer, 0, 4)).order(ByteOrder.LITTLE_ENDIAN).getInt();
            } else {
                Log.e(TAG, "Error in reading byte");
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
        return -1;
    }

    public boolean waitForData() {
        return false;
    }

    public int read(byte[] dest, int bytesToRead) throws IOException {
        int numBytesRead = commonRead(bytesToRead);
        for (int i = 0; i < bytesToRead; i++) {
            dest[i] = buffer[i];
        }
        if (numBytesRead == bytesToRead) {
            return numBytesRead;
        } else {
            Log.e(TAG, "Error in packetHandler Reading");
        }
        return -1;
    }

    public byte[] sendBurst() {
        try {
            commonWrite(burstBuffer.array());
            burstBuffer.clear();
            loadBurst = false;
            int bytesRead = commonRead(inputQueueSize);
            inputQueueSize = 0;
            return Arrays.copyOfRange(buffer, 0, bytesRead);
        } catch (IOException e) {
            e.printStackTrace();
        }
        return new byte[]{-1};
    }

    private int commonRead(int bytesToRead) throws IOException {
        final int[] bytesRead = {0};
        if (mCommunicationHandler.isConnected()) {
            bytesRead[0] = mCommunicationHandler.read(buffer, bytesToRead, timeout);
        } else if (ScienceLabCommon.isWifiConnected()) {
            httpAsyncTask = new HttpAsyncTask(ScienceLabCommon.getEspIP(), new HttpCallback<JSONObject>() {
                @Override
                public void success(JSONObject jsonObject) {
                    try {
                        //Server will send byte array
                        buffer = (byte[]) jsonObject.get("data");
                        bytesRead[0] = buffer.length;
                    } catch (JSONException e) {
                        e.printStackTrace();
                    }
                }

                @Override
                public void error(Exception e) {
                    Log.e(TAG, "Error reading data over ESP");
                }
            });
            httpAsyncTask.execute(new byte[]{});
        }
        return bytesRead[0];
    }

    private void commonWrite(byte[] data) throws IOException {
        if (mCommunicationHandler.isConnected()) {
            mCommunicationHandler.write(data, timeout);
        } else if (ScienceLabCommon.isWifiConnected()) {
            httpAsyncTask = new HttpAsyncTask(ScienceLabCommon.getEspIP(), new HttpCallback<JSONObject>() {
                @Override
                public void success(JSONObject jsonObject) {
                    Log.v(TAG, "write response:" + jsonObject.toString());
                }

                @Override
                public void error(Exception e) {
                    Log.e(TAG, "Error writing data over ESP");
                }
            });

            httpAsyncTask.execute(data);
        }

    }
}

File: lib/communication/packet_handler.dart
import 'dart:convert';
import 'package:flutter/foundation.dart';
import 'package:pslab/communication/commands_proto.dart';
import 'package:pslab/communication/handler/base.dart';
import 'package:pslab/others/logger_service.dart';

class PacketHandler {
  late Uint8List _buffer;
  late CommunicationHandler _mCommunicationHandler;
  static String version = '';
  late CommandsProto _mCommandsProto;
  int _timeout = 500, versionStringLength = 8, fwVersionLength = 3;

  PacketHandler(int timeout, CommunicationHandler communicationHandler) {
    _timeout = timeout;
    _mCommandsProto = CommandsProto();
    _mCommunicationHandler = communicationHandler;
    _buffer = Uint8List(10000);
  }

  bool isConnected() {
    return _mCommunicationHandler.isConnected();
  }

  Future<String> getVersion() async {
    try {
      String scpiResponse = await queryScpi("*IDN?");
      if (scpiResponse.contains("PSLab Pico") ||
          scpiResponse.contains("PSLab Mini")) {
        version = scpiResponse;
        return version;
      }
      sendByte(_mCommandsProto.common);
      sendByte(_mCommandsProto.getVersion);
      await _commonRead(versionStringLength + 1);
      version = utf8
          .decode(_buffer.sublist(0, versionStringLength + 1))
          .split('\n')
          .first;
    } catch (e) {
      logger.e("Error in getting version: $e");
    }

    return version;
  }

  void sendByte(int val) {
    if (!isConnected()) {
      throw Exception("Device not connected");
    }
    try {
      _commonWrite(Uint8List.fromList([val & 0xFF]));
    } catch (e) {
      logger.e("Error in sending byte: $e");
    }
  }

  void sendInt(int val) {
    if (!isConnected()) {
      throw Exception("Device not connected");
    }
    try {
      _commonWrite(Uint8List.fromList([val & 0xFF, (val >> 8) & 0xFF]));
    } catch (e) {
      logger.e("Error in sending int: $e");
    }
  }

  Future<int> getAcknowledgement() async {
    try {
      await _commonRead(1);
      return _buffer[0];
    } catch (e) {
      logger.e(e);
      return 3;
    }
  }

  Future<int> getByte() async {
    try {
      int numBytesRead = await _commonRead(3);
      if (numBytesRead == 3) {
        return _buffer[0];
      } else {
        logger.e("Error in getting voltage");
      }
    } catch (e) {
      logger.e(e);
    }
    return -1;
  }

  Future<int> getVoltageSummation() async {
    try {
      int numBytesRead = await _commonRead(3);
      if (numBytesRead == 3) {
        return (_buffer[0] & 0xFF | ((_buffer[1] << 8) & 0xFF00));
      } else {
        logger.e("Error in getting voltage");
      }
    } catch (e) {
      logger.e(e);
    }
    return -1;
  }

  Future<int> getInt() async {
    try {
      int numBytesRead = await _commonRead(2);
      if (numBytesRead == 2) {
        return (_buffer[0] & 0xFF | ((_buffer[1] << 8) & 0xFF00));
      } else {
        logger.e("Error in reading Int");
      }
    } catch (e) {
      logger.e(e);
    }
    return -1;
  }

  Future<int> getLong() async {
    try {
      int numBytesRead = await _commonRead(4);
      if (numBytesRead == 4) {
        return _buffer.buffer.asByteData(0, 4).getInt32(0, Endian.little);
      } else {
        logger.e("Error in reading Long");
      }
    } catch (e) {
      logger.e(e);
    }
    return -1;
  }

  Future<int> getFirmwareVersion() async {
    try {
      if (version.contains("Pico") || version.contains("Mini")) {
        return 3;
      }
      sendByte(_mCommandsProto.common);
      sendByte(_mCommandsProto.getFwVersion);
      int numBytesRead = await _commonRead(fwVersionLength);
      if (numBytesRead == 1) {
        return 2;
      } else {
        return _buffer[0];
      }
    } catch (e) {
      logger.e(e);
    }
    return 0;
  }

  Future<int> read(Uint8List dest, int bytesToRead) async {
    int numBytesRead = await _commonRead(bytesToRead);

    if (numBytesRead == 0) {
      return 0;
    }
    for (int i = 0; i < numBytesRead; i++) {
      dest[i] = _buffer[i];
    }

    if (numBytesRead == bytesToRead) {
      return numBytesRead;
    } else {
      logger.e(
          "Error in PacketHandler Reading. Expected: $bytesToRead, Got: $numBytesRead");
    }
    return -1;
  }

  Future<int> _commonRead(int bytesToRead) async {
    if (_mCommunicationHandler.isConnected()) {
      return await _mCommunicationHandler.read(_buffer, bytesToRead, _timeout);
    }
    return 0;
  }

  void _commonWrite(Uint8List data) {
    if (_mCommunicationHandler.isConnected()) {
      _mCommunicationHandler.write(data, _timeout);
    }
  }

  Future<String> queryScpi(String command) async {
    String fullCommand = "$command\r\n";
    _mCommunicationHandler.write(
        Uint8List.fromList(fullCommand.codeUnits), 100);
    Uint8List buffer = Uint8List(256);
    int bytesRead = await _mCommunicationHandler.read(buffer, 256, 500);
    if (bytesRead > 0) {
      String response =
          String.fromCharCodes(buffer.sublist(0, bytesRead)).trim();
      return response;
    }
    return "";
  }
}
